Miscellaneous adjustments

rossjrw opened the pull request:
Headers here correspond to headers in the Confluence document detailing my analysis of this interwiki: https://scuttle.atlassian.net/wiki/spaces/TT/pages/1498677249/Interwiki+Rework

Do not merge until #2 has been merged (and then change the base branch to main)

## Misc TODOs

- [x] The script in `styleFrame.html` used to just take the entire query string and treat it as CSS. Need to undo that assumption, as that is now the `css` parameter, and there are other paraeters too.
- [x] Put all files into strict mode.
- [x] Set up a basic ESLint config with at least the essentials ruleset.
- [x] Move all the functions I dumped into `interwiki.js` into more specific files, as appropriate.
- [x] Reduce global variables usage by passing as much info as possible via arguments.
- [x] Update CSS-encoding documentation to be URI-encoded, not percent-encoded - this encoder works fine: https://meyerweb.com/eric/tools/dencoder/; this one doesn't: https://www.url-encode-decode.com/ (compatibility with decodeURIComponent)
- [x] Rewrite the README intro to not imply that this is made for EN
- [x] Test it (obviously)
- [x] Rewrite createResizeIframe
- [ ] Remove `iframe.scpnet-interwiki-frame { height: 350px }` in Sigma-9 (totally optional tbh, because Wikidot will override the height - but may as well set it to 0)
- [x] Make relative theme imports relative to the wiki of the branch set by the lang parameter, not to the site that the iframe was uploaded to
- [ ] Fix resizing (sometimes a scrollbar appears anyway, indicating that the resize was not successful/fired too early; also resizing the iframe after adding each link is too often)
- [x] Remove the link from the header (it looks ugly) - just put the click handler on the p (but also give it a pointer cursor on hover)
- [x] Find a clever way of getting Wikidot base styling into the thing (maybe hardcoding it is reasonable, with a flag to disable it?), and document if necessary

## Supplementary iframes

- [x] Work out what each of them are for

## Additional iframes executing builtin code

- [ ] Put styling for BHL Dark on the BHL Dark Sidebar component.
- [ ] Investigate getting the remaining styles from source as well. (e.g. the default styling for Sigma-9 would ideally be stored alongside Sigma-9's source)

## Tracking added CSS

- [x] Deprecate `styleAry` and give each CSS string its own style element.

## interwikiFrame URL parameters

- [x] Add new interwikiFrame parameters.
- [x] Document all `interwikiFrame.html` parameters, including the new ones.

## Differences between SCP-CN and WL implementations

- [x] Port across more recent WL branches configuration. (#4)
- [ ] Investigate why the WL interwiki allows `PageLookupQModule` to be set to a different module.

## `siteNum`

- [x] ~~Leave `siteNum` for now, but attempt to make it synchronous once I've cleaned everything else up.~~ I've ended up just removing this anyway while cleaning up the function, but I can always recover it if it turns out to be necessary.

## Style containers

- [x] Deprecate the base style (`inStyle`) in favour of a style with priority 0.
- [x] `cuStyle` will become several style elements.

## Exact match for fullnames near the truncation limit

- [x] Rework the truncation detector.
- [x] I forgot to account for when the page has been truncated but the category is removed, resulting in a truncated fullname that would not have been truncated without the category, which would obviously not have any exact matches - detect this condition and account for it

## Knowing which community we're on

- [x] Add a URL parameter for setting the community.
- [x] Document a way to determine the community dynamically via ListPages.

### Making the iframes wait for each other

- [x] Make the styleFrame wait for the interwikiFrame to finish loading before doing its thing.
- [x] ~~Remove~~ Rewrite changeStyleCheck.

## Minimising update effort

- [x] Move everything superfluous out of the HTML files (`interwikiFrame.html` for example should get the query parameters and nothing more, and then immediately defer to a JS script)

## Execution order

- [x] Set up a priority system, determined by URL parameter to styleFrame, to determine the ordering of styles.
- [x] Raise a warning if there are two styles with the same priority.

## New URL parameters for styleFrames

- [x] Add new styleFrame parameters.
- [x] Document them.

rossjrw committed:
Rename 'branches-info' to 'branches-info-scp'

This is consistent with 'branches-info-wl' for the Wanderers' Library,
and makes both filenames explicit.

rossjrw committed:
Install eslint for ES5

rossjrw committed:
Remove ES6 syntax

rossjrw committed:
Use more verbose names for config variables

rossjrw committed:
Rewrite query parameter getter

rossjrw committed:
Start extracting functionality from HTML to script

rossjrw committed:
Move remaining functions to script

rossjrw committed:
Update repository name

rossjrw committed:
Add prospective documentation for styleFrame.html

rossjrw committed:
Get interwiki header by ID

Working with IDs for element selection is just easier.

The header class is retained for CSS compatibility.

rossjrw committed:
Rewrite getDataAll

rossjrw committed:
Rewrite createItem (-> createNewMenuItem)

rossjrw committed:
Rewrite Wikidot pages lookup

rossjrw committed:
Add type definition for Branch

rossjrw committed:
Rewrite function for finding matching pages

getData was the function that looked up a page in a wiki and, if it
found a page with a matching fullname, appended it to the interwiki. I
have renamed it to addTranslationForBranch. getDataAll is now
addTranslations for consistency.

The page category passed to interwikiFrame.html via ListPages on the
sidebar is no longer immediately stripped if it matches the branch's
static category - it is retained, and removed only when the page name
without the category is needed.

I believe my procedure for checking if a matching page exists on another
wiki despite category differences is sound, but it needs testing.
Crucially, the category of the current page will only be stripped if it
matches the current branch's category (i.e. "wanderers:" on an SCP site
without a dedicated WL wiki), and a category for the search will only be
prepended if there is a static category defined in the search candidate
branch's config.

This means for example that page on a dedicated WL wiki that are in a
non-default category will never match any page on a WL branch that does
not have a dedicated wiki - will need to make sure whether this is
desirable.

One issue that still needs solving is how to handle the wanderers-adult
category, if any such category still exists.

rossjrw committed:
Extract lang and community detection to parameters

Community detection (SCP vs WL) was previously performed by checking the
page category, assuming that the 'wanderers' category meant that the
page was WL and everything else was SCP. This was accurate for some
wikis that hosted both SCP and WL articles, but not for dedicated WL
sites.

Instead of attempting to determine the community in code, I have chosen
to leave that responsibility to the person implementing the Interwiki,
who will be much more accurate. The flag has been extracted to a URL
variable. This will also make it easier to test.

The current branch's language code has also been extracted to a URL
parameter. These two parameters will be the only data points needed to
determine which branch the current Interwiki is operating on.

rossjrw committed:
Document interwikiFrame URL parameters and usage

rossjrw committed:
Document new styleFrame query parameters

rossjrw committed:
Recover check for truncated fullnames

There was a 'shorten' check that happened if the target fullname was at
least 59 characters in length, which only took effect after the results
of the page listing had been returned, that confused me and felt
unnecessary, so I removed it. I thought it was supposed to be checking
for differences in fullname as a result of adding a category, and was
doing so incorrectly. I was correct in that this feature was missing,
and rightfully added it. However, I was wrong in what the 'shorten' flag
was actually checking for - it was actually handling the case for when a
category had been removed. I have recovered this check.

rossjrw committed:
Handle internal and external styles separately

I've moved all of the style-handling functionality out into another
script.

The 'theme' parameter of interwikiFrame enables it to accept a link to a
full CSS stylesheet stored somewhere else. While it's theoretically
equivalent to sending an @import to the 'css' parameter, there's no
reason not to coerce that value into a <link> element, which offers some
performance benefits over <style>.

I've rewritten the function that handles a style change, and then
additionally split it up into two functions, each handling internal
(<style>) and external (<link>) styles respectively.

I've also added the priority sorting system, which sorts all the styling
elements against a priority value. I've stated elsewhere that the 'css'
parameter will override a 'theme' parameter if added in the same
styleFrame (and therefore they have the same priority); to account for
this, link elements have an infinitesimally lower priority than style
elements in the sorting algorithm, implemented via a single conditional
check.

I have also added a requestStyleChange function that accepts the
location.search value of a styleFrame directly. This function will be
called from styleFrames to avoid them having to parse their own
arguments (as it's important that both HTML files contain as little
logic as possible), and it will also be called from the new
changeStyleCheck().

rossjrw committed:
Get query parameters from specified location

As opposed to getting query parameters only from the location of the
current document. This is necessary in order to be able to extract query
parameters from the location of the styleFrame.

rossjrw committed:
Rewrite styleFrame

rossjrw committed:
Remove builtin style handlers

The builtin styles will be moved to wherever the source for their
respective styles is. There is no reason to keep hardcoded copies of
those styles.

The actual CSS files representing these styles will be retained until
they've been ported to their appropriate locations, however.

rossjrw committed:
Move translation link related functions to file

rossjrw committed:
Rewrite changeStyleCheck (-> pullStyles)

rossjrw committed:
Extract most JS in interwiki HTML to file

This minimises the chance that the HTML file will need to be updated
after any given update to the Interwiki.

rossjrw committed:
Add ESLint recommended config and make compliant

rossjrw committed:
Enforce strict mode

rossjrw committed:
Rewrite createResizeIframe

The procedure to resize the iframe was previously being called every
50ms, forever. This is not an ideal way to go about things.

The updated version now constructs a function to resize the iframe,
which is more consistent with its name. This function will be
constructed in the main interwiki script (as opposed to in parallel with
it), and will be passed to the translation generator. It will then be
called after adding each new translation, ensuring that it is only ever
called as needed and never superfluously.

rossjrw committed:
Resize the iframe after adding a translation link

rossjrw committed:
Add doctype

rossjrw committed:
Fix parent URL getter

rossjrw committed:
Fix the resizing process

rossjrw committed:
Hide the empty side block by default

rossjrw committed:
Construct header link in JS

rossjrw committed:
Promote priority conflict warning to error

Wikidot emits enough warnings already that another one won't be noticed.
It's a soft error, not a hard error, so it's fine.

rossjrw committed:
Document CSS styling

rossjrw committed:
Use protocol-relative URLs for iframes

rossjrw committed:
Put HTML files into strict mode

Also move the isInterwikiFrame exposure to run at the end of the
interwikiFrame init, so that styleFrame doesn't prematurely inject its
style.

rossjrw committed:
Fix translation link ordering

rossjrw committed:
Rewrite README intro to be less SCP EN-centric

rossjrw committed:
Fix documentation for iframe resize wrapper

emmiegit commented on js/interwiki.js:
```diff
@@ -0,0 +1,84 @@
+ * @param {String} query - The URL query.
+ * @param {String} name - The name of the parameter to get.
+ */
+function getQueryString(query, name) {
```
Instead of re-parsing it every time, I would have this function extract all the arguments into a map, then you just refer to that. You can preserve "later values override earlier ones" by just iterating in-order.

rossjrw replied:
A given query string shouldn't have more than 3 parameters, so I don't expect the performance hit of reparsing to even be measurable. I'm sure there's a much saner and robust URLSearchParams polyfill out there somewhere.

emmiegit commented on js/links.js:
```diff
@@ -0,0 +1,242 @@
+      try {
+        if (request.status === 200) {
+          var response = JSON.parse(request.responseText);
+          // {"pages":[{"unix_name":"scp-xxx","title":"SCP-XXX"}]}
```
I would include a "format:" or similar so it doesn't look like commented-out code

rossjrw replied:
f378899

emmiegit commented on js/links.js:
```diff
@@ -0,0 +1,242 @@
+          });
+        }
+      } catch (error) {
+        // Parsing failed - assume there are no matching pages
```
Can we at least `console.error()` here?

rossjrw replied:
32e9b7d

emmiegit commented on js/styles.js:
```diff
@@ -0,0 +1,197 @@
+ */
+function urlFromTheme(theme) {
+  // If the theme is already a full URL, return it
+  if (theme.indexOf("http") === 0 || theme.indexOf("//") === 0) return theme;
```
Having single-line `if`s go from inline to next line is confusing imo

rossjrw replied:
Fair enough - my rationale is that <code>return theme</code> is practically a no-op so an inline return sort of makes sense, but the next return concats some strings so it's not a no-op so an inline return makes less sense, even though it would fit on the line. But I agree in retrospect that it does look a bit weird like this. fdb3d1c

rossjrw committed:
Tag API response format sample as such

rossjrw committed:
Report lookup errors to console

rossjrw committed:
Use multi-line condition for consistency

rossjrw committed:
Document theme parameters options more explicitly

rossjrw committed:
Use site URL to construct relative theme URLs

A relative URL from the perspective of the interwikiFrame resolves to
the wdfiles domain that the iframe was uploaded to. The intended use
case is for all sites to use the same HTML file, which is uploaded to a
single site in the network, so all relative styles would resolve to that
site which is obviously not tenable. (This is also why failed calls to
urlFromTheme return the empty string rather than a domain-relative URL,
as those URLs would work on one wiki and issues would go unnoticed, but
would not work on other wikis.)

This commits changes the function to accept a siteUrl argument.

rossjrw committed:
Add factory for restyle function

requestStyleChange (which is called from styleFrame and also used
internally by interwikiFrame) used to have a single argument, which was
the request to parse. However, after the previous commit it is now
required to also pass the interwiki's configured base URL to the theme
URL constructor. The styleFrame cannot possibly know this information.

Instead of defining the function globally, it is now constructed during
the interwiki initialisation by createRequestStyleChange which takes the
needed URL as an argument. This is exposed to the window, so it is
effectively a global function.

rossjrw committed:
Do not catch all errors when restyling

Exactly half of restyle requests are intended to fail, so any errors
emitted from them were caught and discarded. However, this naive
approach prevents actual errors from being reported.

The error handler has been changed to rethrow if the error is not a
DOMException, which is the type of error that is expected.

rossjrw committed:
Insert low-priority styles correctly

rossjrw committed:
Add Wikidot base styling by default

I'm not totally comfortable with hardcoding any CSS into the new
interwiki, even this, but because I can guarantee that this is going to
be needed by all sites that this interwiki intends to support I think
it's a pretty reasonable addition.

When there comes a time that a site does not need the base Wikidot
styling for whatever reason - maybe a site on Wikijump, or a site hosted
elsewhere - there are other changes that would need to be made anyway.
So it's probably fine.

But just to be safe, I added a parameter to turn the automatic styling
off, which I'm not going to document.

rossjrw committed:
Remove link from header

Click-to-refresh is now activated by just clicking the header. Not
putting a link there makes the appearance of the interwiki consistent
with other sidebar blocks without needing special styling. The
click-to-refresh functionality isn't important to end users, so it
doesn't matter that it's not clearly broadcasted.

rossjrw committed:
Debounce the resizer

rossjrw committed:
Resize iframe after adding new styles

There is no way to tell when CSS from a link or style element has been
applied that I could find. There is a load event, and there is a
polyfill for the load event that uses the error event on images, but
both of these events fire before the CSS has actually been applied to
the document. I brute-forced it - I added a timeout.

The resizer is already debounced, so adding frivolous resizes is not a
performance concern.

rossjrw committed:
Avoid adding padding/margin to html/body
